src/libsyntax/tokenstream.rs

```diff
@@ -156,15 +160,22 @@ impl TokenTree {
 #[derive(Clone, Debug)]
 pub enum TokenStream {
     Empty,
-    Tree(TokenTree),
-    JointTree(TokenTree),
+    Tree(TokenTree, IsJoint),
     Stream(Lrc<Vec<TokenStream>>),
 }
 
 // `TokenStream` is used a lot. Make sure it doesn't unintentionally get bigger.
 #[cfg(target_arch = "x86_64")]
 static_assert!(MEM_SIZE_OF_TOKEN_STREAM: mem::size_of::<TokenStream>() == 32);
 
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum IsJoint {
+    Joint,
+    NonJoint
+}
```

Comment:
Is it possible to reuse `proc_macro::Spacing` directly instead of introducing a new enum?

Reply:
It would be nice to make compiler-internal token stream structures as close to `proc_macro` ones as possible (modulo performance tweaks, perhaps).

Reply:
I'm not sure `proc_macro::Spacing` has the right "semantics" here - the compiler's `TokenStream` doesn't preserve such information itself, and this "jointness" is only enabled by conversion from `proc_macro::Punct`.

This, in turn, then causes tokens to be joined on the fly (as the compiler has "pre-joined" tokens, e.g. `DotDotEq` for `..=`).

Ideally we'd align the compiler more with the proc macro API, but things like e.g. `macro_rules` may have specific expectations based on those "pre-joined" tokens existing.

Reply:
Based on the above, I will keep `IsJoint`.